### PR TITLE
Zipkin tracing

### PR DESCRIPTION
**Pull request by oschaaf**
Opening an early PR for those who would like to have a look-see at progress around adding OpenTracing capabilities in Nighthawk. The current status is that this carves a crude but working path towards end-to-end tracing. 

Tasks left:
- Inventorize what information is missing in the NH part of the trace and add it
- Generalize set up of the tracing in NH so we can have other implementations other then zipkin

To keep the PR small, maybe it's a good idea to land a minimal version first.

Example record in the zipkin UI:

image

ps - At the moment the diff for this PR is rather large, but that will become a lot smaller once we land the outstanding PR's and merge master after that.

/cc @htuch

### CODE REVIEW

File: include/nighthawk/client/factories.h [hunk 1]
```diff
@@ -25,8 +25,9 @@ class BenchmarkClientFactory {
 public:
   virtual ~BenchmarkClientFactory() = default;
   virtual BenchmarkClientPtr create(Envoy::Api::Api& api, Envoy::Event::Dispatcher& dispatcher,
-                                    Envoy::Stats::Scope& scope,
```
Comment (jplevyak) [inline include/nighthawk/client/factories.h]: Why rename this?

File: source/client/process_impl.cc [hunk 2]
```diff
@@ -245,17 +246,71 @@ ProcessImpl::createBootstrapConfiguration(const Uri& uri, int number_of_clusters
...
+  config.mutable_shared_span_context()->set_value(true);
+  http->mutable_typed_config()->PackFrom(config);
+}
 
-  return bootstrap;
+void ProcessImpl::maybeCreateTracingDriver(const envoy::config::trace::v2::Tracing& configuration) {
+  if (configuration.has_http()) {
+    std::string type = configuration.http().name();
+    ENVOY_LOG(info, "loading tracing driver: {}", type);
+    // Envoy::Server::Configuration::TracerFactory would be useful here to create the right
```
Comment (jplevyak) [inline source/client/process_impl.cc]: Is this comment still accurate?

File: source/client/stream_decoder.cc [hunk 3]
```diff
@@ -4,60 +4,83 @@
...
   if (complete_) {
     onComplete(true);
   }
 }
 
-void StreamDecoder::decodeTrailers(Envoy::Http::HeaderMapPtr&&) {
+void StreamDecoder::decodeTrailers(Envoy::Http::HeaderMapPtr&& headers) {
   ASSERT(!complete_);
   complete_ = true;
+  trailer_headers_ = std::move(headers);
```
Comment (jplevyak) [inline source/client/stream_decoder.cc]: Comment why we need the headers?

File: source/client/stream_decoder.cc [hunk 4]
```diff
@@ -4,60 +4,83 @@
...
 void StreamDecoder::onPoolReady(Envoy::Http::StreamEncoder& encoder,
                                 Envoy::Upstream::HostDescriptionConstSharedPtr) {
+  upstream_timing_.onFirstUpstreamTxByteSent(time_source_); // XXX(oschaaf): is this correct?
   encoder.encodeHeaders(*request_headers_, request_body_size_ == 0);
   if (request_body_size_ > 0) {
+    // This will show up in the zipkin UI as 'response_size'.
+    // We add it here, optimistically assuming it will all be send.
+    // Ideally, we'd track the encoder events of the stream to dig up and forward
+    // more information. For now, we take the risk of erroneously reporting that
+    // we did send all the bytes, instead of always reporting 0 bytes.
```
Comment (jplevyak) [inline source/client/stream_decoder.cc]: Do you want to create an issue to fix this (w/TODO)?

File: source/client/stream_decoder.cc [hunk 5]
```diff
@@ -71,5 +94,34 @@ void StreamDecoder::onPoolReady(Envoy::Http::StreamEncoder& encoder,
   }
 }
 
+// TODO(oschaaf): duplicated from the envoy code base.
+// See if it can be moved to make it more easily shareable there, and drop it here.
+Envoy::StreamInfo::ResponseFlag
+StreamDecoder::streamResetReasonToResponseFlag(Envoy::Http::StreamResetReason reset_reason) {
```
Comment (jplevyak) [inline source/client/stream_decoder.cc]: Do you want an issue for this?

File: source/client/stream_decoder.cc [hunk 6]
```diff
@@ -71,5 +94,34 @@ void StreamDecoder::onPoolReady(Envoy::Http::StreamEncoder& encoder,
...
+    return Envoy::StreamInfo::ResponseFlag::UpstreamOverflow;
+  case Envoy::Http::StreamResetReason::RemoteReset:
+  case Envoy::Http::StreamResetReason::RemoteRefusedStreamReset:
+    return Envoy::StreamInfo::ResponseFlag::UpstreamRemoteReset;
+  }
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+void StreamDecoder::finalizeActiveSpan() {
+  if (active_span_.get() != nullptr) {
```
Comment (jplevyak) [inline source/client/stream_decoder.cc]: you don't have to add the .get() here.

File: source/client/stream_decoder.h [hunk 7]
```diff
@@ -35,14 +41,28 @@ class StreamDecoder : public Envoy::Http::StreamDecoder,
...
       : dispatcher_(dispatcher), time_source_(time_source),
         decoder_completion_callback_(decoder_completion_callback),
         caller_completion_callback_(std::move(caller_completion_callback)),
         connect_statistic_(connect_statistic), latency_statistic_(latency_statistic),
         request_headers_(std::move(request_headers)), connect_start_(time_source_.monotonicTime()),
         complete_(false), measure_latencies_(measure_latencies),
-        request_body_size_(request_body_size) {}
+        request_body_size_(request_body_size), stream_info_(time_source_),
+        http_tracer_(http_tracer) {
+    if (measure_latencies_ && http_tracer_ != nullptr) {
```
Comment (jplevyak) [inline source/client/stream_decoder.h]: This is a lot for a constructor... is the a better place to do this?

File: test/client_test.cc [hunk 8]
```diff
@@ -42,6 +42,29 @@ TEST_F(ClientTest, AutoConcurrencyRun) {
   EXPECT_TRUE(program.run());
 }
 
+// TODO(oschaaf): this is just for coverage, and we don't care where
+// any traffic we send ends it or what that looks like.
+// We do functional testing in python now, but unfortunately any code we hit
+// there isn't counted as code-coverage.
+// Ideally, the code hit during the python test runs would count for covergage,
+// and we use unit-tests here to hit any edge cases we can't easily hit otherwise.
+// XXX(oschaaf): look into ways to accomplish that.
```
Comment (jplevyak) [inline test/client_test.cc]: Do you want an issue?  I like the style where you do TODO(issue URI).  It makes it easy to cross-reference the issue and the code and it doesn't tie the work to one person.

File: test/integration/test_integration_basics.py [hunk 9]
```diff
@@ -272,3 +274,22 @@ def test_https_log_verbosity(https_test_server_fixture):
...
   assertIn(trace_level_sentinel, logs)
+
+
+def test_tracing_zipkin(http_test_server_fixture):
+  """
+  Test that we send spans when our zipkin tracing feature
+  is enabled. Note there's no actual zipkin server started, so
+  traffic will get (hopefully) get send into the void.
+  """
+  # TODO(oschaaf): boot up an actual zipkin server to accept
```
Comment (jplevyak) [inline test/integration/test_integration_basics.py]: Issue?

File: test/process_test.cc [hunk 10]
```diff
@@ -26,11 +26,11 @@ class ProcessTest : public TestWithParam<Envoy::Network::Address::IpVersion> {
       : loopback_address_(Envoy::Network::Test::getLoopbackAddressUrlString(GetParam())),
         options_(TestUtility::createOptionsImpl(
             fmt::format("foo --duration 1 -v error --rps 10 https://{}/", loopback_address_))){};
-  void runProcess() {
+  void runProcess(bool expect_success = true) {
```
Comment (jplevyak) [inline test/process_test.cc]: Consider doing enum class { EXPECT_SUCCESS, EXPECT_FAILURE} which makes the code easier to read without comments.

File: test/stream_decoder_test.cc [hunk 11]
```diff
@@ -41,14 +42,16 @@ class StreamDecoderTest : public Test, public StreamDecoderCompletionCallback {
...
 };
 
 TEST_F(StreamDecoderTest, HeaderOnlyTest) {
   bool is_complete = false;
   auto decoder = new StreamDecoder(
       *dispatcher_, time_system_, *this, [&is_complete](bool, bool) { is_complete = true; },
-      connect_statistic_, latency_statistic_, request_headers_, false, 0);
-  auto headers = std::make_unique<Envoy::Http::HeaderMapImpl>();
+      connect_statistic_, latency_statistic_, request_headers_, false, 0, "654", http_tracer_);
+  Envoy::Http::HeaderMapPtr headers{new Envoy::Http::TestHeaderMapImpl{{":status", "200"}}};
```
Comment (jplevyak) [inline test/stream_decoder_test.cc]: consider auto headers = std::make_X

Comment (jplevyak) [reply test/stream_decoder_test.cc]: and below

File: test/stream_decoder_test.cc [hunk 12]
```diff
@@ -41,14 +42,16 @@ class StreamDecoderTest : public Test, public StreamDecoderCompletionCallback {
   HeaderMapPtr request_headers_;
   uint64_t stream_decoder_completion_callbacks_{0};
   uint64_t pool_failures_{0};
+  // XXX(oschaaf): mock tracer, set expectations
```
Comment (oschaaf) [inline test/stream_decoder_test.cc]: Mock tracer added + expectations added here.
